include/zephyr/toolchain/gcc.h

```diff
@@ -64,6 +64,7 @@
 #endif
 
 
+#undef BUILD_ASSERT /* clear out common version */
```

Comment:
> BUILD_ASSERT is also defined in include/toolchain/common.h, which might
get included before gcc.h. We want to use the gcc-specific one instead
of the general one.

`toolchain/common.h` is never supposed to be included before the toolchain-specific headers such as `toolchain/gcc.h`; in fact, it should not be included by any files other than the toolchain-specific headers inside `toolchain/`.

Properly documenting and enforcing that seems to be beyond the scope of this PR, however.